src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,246 @@
+    return r;
+  }
+  return r;
+}
+
+} // anonymous namespace
+
+int migrate(const DoutPrefixProvider* dpp,
+            rgw::sal::RadosStore* driver,
+            boost::asio::io_context& context,
+            spawn::yield_context yield)
+{
+  auto y = optional_yield{context, yield};
+
+  ldpp_dout(dpp, 1) << "starting v1 topic migration.." << dendl;
+
+  librados::Rados* rados = driver->getRados()->get_rados_handle();
+  const rgw_pool& pool = driver->svc()->zone->get_zone_params().log_pool;
+  librados::IoCtx ioctx;
+  int r = rgw_init_ioctx(dpp, rados, pool, ioctx);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to initialize log pool for listing with: "
+        << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // loop over all objects with oid prefix "pubsub."
+  auto filter = rgw::AccessListFilterPrefix(rgw::sal::pubsub_oid_prefix);
+  constexpr uint32_t max = 100;
+  std::string marker;
+  bool truncated = false;
+
+  std::vector<std::string> oids;
+  do {
+    oids.clear();
+    r = rgw_list_pool(dpp, ioctx, max, filter, marker, &oids, &truncated);
+    if (r == -ENOENT) {
+      r = 0;
+      break;
+    }
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to list v1 topic metadata with: "
+          << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    for (const std::string& oid : oids) {
+      const auto obj = rgw_raw_obj{pool, oid};
+      if (oid.find(rgw::sal::pubsub_bucket_oid_infix) != oid.npos) {
+        ldpp_dout(dpp, 4) << "migrating v1 topics " << oid << dendl;
+        r = migrate_notification(dpp, y, driver, obj);
+        ldpp_dout(dpp, 4) << "migrating v1 topics " << oid << " completed with: "
+                          << ((r == 0)? "successful": cpp_strerror(r)) << dendl;
+      } else {
+        ldpp_dout(dpp, 4) << "migrating v1 bucket topics " << oid << dendl;
+        r = migrate_topics(dpp, y, driver, obj);
+        ldpp_dout(dpp, 4) << "migrating v1 bucket topics " << oid << " completed with: "
+                          << ((r == 0)? "successful": cpp_strerror(r)) << dendl;
+      }
+    }
+    if (!oids.empty()) {
+      marker = oids.back(); // update marker for next listing
+    }
+  } while (truncated);
+
+  ldpp_dout(dpp, 1) << "finished v1 topic migration" << dendl;
+  return 0;
+}
+
+int deconstruct_topics_oid(const std::string& bucket_topics_oid, std::string& tenant, std::string& bucket_name,
+                                        std::string& marker, const DoutPrefixProvider* dpp) {
+  auto pos = bucket_topics_oid.find(rgw::sal::pubsub_bucket_oid_infix);
+  if (pos == std::string::npos) {
+    ldpp_dout(dpp, 1) << "ERROR: bucket_topics_oid:" << bucket_topics_oid << " doesn't contain " << rgw::sal::pubsub_bucket_oid_infix
+                      << " after tenant name!" << dendl;
+    return -EINVAL;
+  }
+  tenant = bucket_topics_oid.substr(0, pos);
+
+  auto bucket_name_marker = bucket_topics_oid.substr(tenant.size() + rgw::sal::pubsub_bucket_oid_infix.size());
```

Comment:
`pos` is at the beginning of `pubsub_bucket_oid_infix`, so this can be
```c++
auto bucket_name_marker = bucket_topics_oid.substr(pos + rgw::sal::pubsub_bucket_oid_infix.size());
```

Reply:
Thanks  alot :)